objects/src/block/note_tree.rs

```diff
@@ -0,0 +1,138 @@
+pub struct BlockNoteTree(SimpleSmt<BLOCK_OUTPUT_NOTES_TREE_DEPTH>);
+
+impl BlockNoteTree {
+    /// Returns a new [BlockNoteTree] instantiated with entries set as specified by the provided entries.
+    ///
+    /// Entry format: (batch_index, note_index, (note_id, note_metadata)).
+    ///
+    /// All leaves omitted from the entries list are set to [ZERO; 4].
+    ///
+    /// # Errors
+    /// Returns an error if:
+    /// - The number of entries exceeds the maximum notes tree capacity, that is 2^21.
+    /// - The provided entries contain multiple values for the same key.
+    pub fn with_entries(
+        entries: impl IntoIterator<Item = (usize, usize, (RpoDigest, NoteMetadata))>,
+    ) -> Result<Self, MerkleError> {
+        let interleaved =
+            entries.into_iter().flat_map(|(batch_index, note_index, (note_id, metadata))| {
+                let id_index = Self::leaf_index(batch_index, note_index);
+                [(id_index, note_id.into()), (id_index + 1, metadata.into())]
+            });
+
+        SimpleSmt::with_leaves(interleaved).map(Self)
+    }
+
+    /// Returns the root of the tree
+    pub fn root(&self) -> RpoDigest {
+        self.0.root()
+    }
+
+    /// Returns merkle path for the note with specified batch/note indexes
+    pub fn merkle_path(
+        &self,
+        batch_idx: usize,
+        note_idx_in_batch: usize,
+    ) -> Result<MerklePath, MerkleError> {
+        let leaf_index =
+            LeafIndex::<NOTE_LEAF_DEPTH>::new(Self::note_index(batch_idx, note_idx_in_batch))?;
+
+        Ok(self.0.open(&leaf_index).path)
+    }
+
+    // HELPERS
+    // --------------------------------------------------------------------------------------------
+
+    fn note_index(batch_idx: usize, note_idx_in_batch: usize) -> u64 {
+        (batch_idx * MAX_NOTES_PER_BATCH + note_idx_in_batch) as u64
+    }
+
+    fn leaf_index(batch_idx: usize, note_idx_in_batch: usize) -> u64 {
+        Self::note_index(batch_idx, note_idx_in_batch) * 2
+    }
+}
+
+impl Default for BlockNoteTree {
+    fn default() -> Self {
+        Self(SimpleSmt::new().expect("Unreachable"))
+    }
+}
+
+// SERIALIZATION
+// ================================================================================================
+
+const EOF_INDEX: u32 = 0xFFFF_FFFF;
+
+impl Serializable for BlockNoteTree {
+    fn write_into<W: ByteWriter>(&self, target: &mut W) {
+        for (index, node) in self.0.leaves() {
+            target.write_u32(index as u32);
+            target.write(node);
+        }
+        target.write_u32(EOF_INDEX);
+    }
+}
+
+impl Deserializable for BlockNoteTree {
+    fn read_from<R: ByteReader>(source: &mut R) -> Result<Self, DeserializationError> {
+        let mut leaves = vec![];
+        loop {
+            let index = source.read_u32()?;
+            if index == EOF_INDEX {
+                break;
+            }
+            let value: Word = source.read()?;
+            leaves.push((index as u64, value));
+        }
```

Comment:
Question: Why is this using this approach instead of length prefix? Length prefix has a few benefits:

- It is the format used elsewhere, so it would make the code more consistent / easier to follow.
- It allows to skip data by computing offsets

Reply:
@hackaugusto I used this approach because it can help to avoid allocation/calculation of tree length during writing. It can also help to avoid allocation during reading, but this is requires to support results in iterator elements in `SimpleSmt::with_leaves` associated function.

Reply:
> @hackaugusto I used this approach because it can help to avoid allocation/calculation of tree length during writing. It can also help to avoid allocation during reading, but this requires support of results in iterator elements in SimpleSmt::with_leaves associated function.

Why is that? On the write path, is it because you didn't have an API to get the number of non-empty leaves in the `SimpleSMT`? What about adding that API to crypto?

On the reading size, from what I can tell this does the opposite, and causes more allocations to happen, because `leaves.push` will have internal allocations and copies. With a length prefix one can do a `Vec::with_capacity`.

Reply:
@hackaugusto you are right, thank you! I will update the code.

Reply:
@hackaugusto I've migrated miden-base to the latest crypto (it wasn't straightforward, by the way, so it also needs your review or who is the right person to do it?), please take a look!

Reply:
The migration work is also done in https://github.com/0xPolygonMiden/miden-base/pull/561. I'll make a new release of `miden-vm` tomorrow, and then we can merge both of these PRs.

Reply:
@bobbinth I've merged #561 and will merge this PR. Please ping me, once `miden-vm` is released, I will create separate PR for updating dependencies to crates.io.